KEYMGMT: export/import domain parameters

levitte opened the pull request:
Including DH and DSA functionality for import and export of domain parameters and keys

This is the foundation needed to manipulate domain parameters and keys with providers beyond what we currently do.  What we've done so far is allow libcrypto generated keys to be transported to providers to be able to use them for signature and key exchange operations.  However, with things like key generation, not to mention *any* domain parameter handling, we're soon going to find ourselves in a situation where the libcrypto generated keys are not relevant at all for some operations (such as domain parameter and key generation).

Tests will happen as soon as the diverse upcoming PKEY operations start using this functionality.  (next up is key generation)

levitte committed:
PKEY: adapt the export_to_provider funtions to handle domain params too

levitte committed:
DH: Add export of domain parameters to provider

levitte committed:
DSA: Add export of domain parameters to provider

levitte committed:
DH: Add export of keys and domain parameters from provider

levitte committed:
DSA: Add export of keys and domain parameters from provider

mattcaswell commented on crypto/dh/dh_ameth.c:
```diff
@@ -566,34 +567,39 @@ static void *dh_pkey_export_to(const EVP_PKEY *pk, EVP_KEYMGMT *keymgmt)
+        if (q != NULL) {
+            if (!ossl_param_bld_push_BN(&tmpl, OSSL_PKEY_PARAM_FFC_Q, q))
+                return NULL;
+        }
```
`q` is an (optional) domain parameter (not yet supported by our provider implementations)

levitte replied:
Ah, gotcha

mattcaswell commented on crypto/dh/dh_ameth.c:
```diff
@@ -566,34 +567,39 @@ static void *dh_pkey_export_to(const EVP_PKEY *pk, EVP_KEYMGMT *keymgmt)
+         * data - so "pub_key" is optional. We can never have a "priv_key"
+         * without a corresponding "pub_key" though.
+         */
+        if (pub_key != NULL) {
```
So if `domainparams` is set to false, then surely `pub_key` should never be NULL? Its an error if it is.

levitte replied:
From the `EVP_get0_key` manual (Italic bold mine):

> To get the public and private key values use the **DH_get0_key()** function. A pointer to the public key will be stored in **\*pub_key**, and a pointer to the private key will be stored in **\*priv_key**. Either may be NULL if they have not been set yet, although ***if the private key has been set then the public key must be***. The values point to the internal representation of the public key and private key values. This memory should not be freed directly.  Any of the out parameters **pub_key** and **priv_key** can be NULL, in which case no value will be returned for that parameter.

Do you remember that we discussed this once already?

mattcaswell replied:
I'm somehow missing your point?

levitte replied:
Hmmm, so am I.  As a matter of fact, I suggest you take it up with whoever wrote [this PR](https://github.com/openssl/openssl/pull/9753) :wink:, that's where this was introduced

mattcaswell replied:
It makes sense in that PR because keys could be empty except for domain parameters. In this PR you are splitting them up. Therefore it no longer makes sense to send a *key* without at least the pub key set.

levitte replied:
I must be daft, I suddenly see what you see...

levitte replied:
That comment is *really* confusing, in hindsight

mattcaswell commented on crypto/dsa/dsa_ameth.c:
```diff
@@ -552,30 +553,34 @@ static void *dsa_pkey_export_to(const EVP_PKEY *pk, EVP_KEYMGMT *keymgmt)
+         * data - so "pub_key" is optional. We can never have a "priv_key"
+         * without a corresponding "pub_key" though.
+         */
+        if (pub_key != NULL) {
```
As with dh isn't it an error if `pub_key` is NULL?

slontis commented on providers/implementations/keymgmt/dh_kmgmt.c:
```diff
@@ -14,18 +14,68 @@
+
+    if (!DH_set0_pqg(dh, p, NULL, g))
+        goto err;
+    p = g = NULL;
```
Not sure why this line in needed.

levitte replied:
It was there before, and didn't think to clean it out.

levitte replied:
Removed

slontis commented on providers/implementations/keymgmt/dh_kmgmt.c:
```diff
@@ -40,31 +90,65 @@ static int params_to_key(DH *dh, const OSSL_PARAM params[])
-
     if (!DH_set0_key(dh, pub_key, priv_key))
         goto err;
     priv_key = pub_key = NULL;
```
Same here..

levitte replied:
Done

slontis commented on providers/implementations/keymgmt/dsa_kmgmt.c:
```diff
@@ -14,20 +14,74 @@
+
+    if (!DSA_set0_pqg(dsa, p, q, g))
+        goto err;
+    p = q = g = NULL;
```
?

levitte replied:
Removed

slontis commented on providers/implementations/keymgmt/dsa_kmgmt.c:
```diff
@@ -40,34 +94,64 @@ static int params_to_key(DSA *dsa, const OSSL_PARAM params[])
-
     if (pub_key != NULL && !DSA_set0_key(dsa, pub_key, priv_key))
         goto err;
     priv_key = pub_key = NULL;
```
?

levitte replied:
Removed

slontis commented:
looks like 30-test_evp_extra.t doesnt like something in the changes...

levitte commented:
Yeah...  the declarative way to define of the numbers being transferred are just the domain parameters or the full key doesn't quite work, there are many places where we don't make that distinction, while the KEYMGMT interface suggests a separation.

There are two ways to handle that:  have the export functions auto-detect what the given key is, or ditch the distinction.  I'm currently leaning toward the former...

levitte commented:
At the same time...  the test that fails is a bit abusive, it initializes a signature operation with an `EVP_PKEY` that are just the domain parameters.  Is that really what that test should do, check how much abuse the API can tolerate?  ...  and even more, expect success on the abuse?

levitte committed:
Don't abuse the API when that's not what is tested

test_EVP_PKEY_CTX_get_set_params() in test/evp_extra_test.c abused
previously sloppy checking in EVP_PKEY_sign_init_ex(), by passing a
"key to sign with" that was really just domain parameters.

Now that underlying provider import of key payload has become a bit
more strict, that leads to errors, so we need to provide at least a
public part (even though fake), and because this is a signing
operation, a private part as well.

t8m commented:
IMO initialization of a signature operation just with domain parameters and no key should simply fail.

levitte commented:
> IMO initialization of a signature operation just with domain parameters and no key should simply fail.

Yup, that is my conclusion as well.  I've adjusted `test/evp_extra_test.c` accordingly, since `test_EVP_PKEY_CTX_get_set_params()` isn't there to check how much we can abuse the API.

levitte commented:
Distinguishing domain params from full keys wasn't as complicated as I was afraid it was going to be.  The low level export functions could do that pretty easily, the rest was all about matching expectations.

levitte commented:
CIs are happy now!

slontis commented on crypto/dh/dh_ameth.c:
```diff
@@ -565,19 +566,16 @@ static void *dh_pkey_export_to(const EVP_PKEY *pk, EVP_KEYMGMT *keymgmt)
-     */
-    if (pub_key != NULL) {
-        if (!ossl_param_bld_push_BN(&tmpl, OSSL_PKEY_PARAM_DH_PUB_KEY, pub_key))
+    *is_domainparams = (pub_key == NULL && priv_key == NULL);
```
Can domainparams be NULL?

levitte replied:
Never in libcrypto, as far as I've been able to detect

slontis commented on crypto/evp/keymgmt_lib.c:
```diff
@@ -138,7 +152,7 @@ void *evp_keymgmt_export_to_provider(EVP_PKEY *pk, EVP_KEYMGMT *keymgmt)
                 params = paramdefs_to_params(paramdefs);
                 /* Get 'return_size' filled */
-                pk->pkeys[j].keymgmt->exportkey(pk->pkeys[j].provkey, params);
+                exportfn(pk->pkeys[j].provdata, params);
```
why is importfn checked for NULL but exportfn isnt?

levitte replied:
In this particular case, it's all about getting the `return_size` field filled in.  The return value from exportfn is irrelevant here.  However, I've added a check on the next one.

slontis commented on providers/implementations/keymgmt/dsa_kmgmt.c:
```diff
@@ -40,34 +94,64 @@ static int params_to_key(DSA *dsa, const OSSL_PARAM params[])
  err:
-    BN_free(p);
-    BN_free(q);
-    BN_free(g);
```
are p,q,g now freed as part of the dsa free?

levitte replied:
This function is only about pub_key and priv_key, the domain parameters are handled in a separate function.

slontis replied:
You do params_to_domparams() in this function. My question was related to who frees this is an error occurs later in this function

levitte committed:
fixup! DH: Add export of domain parameters to provider

levitte committed:
fixup! DSA: Add export of domain parameters to provider

levitte commented:
CIs are still happy, and I hope I've addressed all review comments.  Anything else?

levitte commented on crypto/evp/keymgmt_lib.c:
```diff
@@ -173,12 +197,25 @@ void *evp_keymgmt_export_to_provider(EVP_PKEY *pk, EVP_KEYMGMT *keymgmt)
+     * other way around isn't as certain.
+     */
+    if (want_domainparams || want_domainparams == is_domainparams)
+        return provdata;
```
Thanks.  I do want exactly that separation that you're talking about, and I thought I got something "smart", but it's obvious i need to look again to make sure.

levitte replied:
OK, I've changed it so the `want_domainparams` is demanding the whole way through instead of doing this discovery thing.  Please have a look.

levitte committed:
fixup! PKEY: adapt the export_to_provider funtions to handle domain params too

levitte committed:
fixup! DH: Add export of domain parameters to provider

levitte committed:
fixup! DSA: Add export of domain parameters to provider

levitte commented:
Ping.

mattcaswell commented on crypto/dh/dh_ameth.c:
```diff
@@ -590,10 +587,12 @@ static void *dh_pkey_export_to(const EVP_PKEY *pk, EVP_KEYMGMT *keymgmt)
-    provkey = evp_keymgmt_importkey(keymgmt, params);
+    provdata = want_domainparams
+        ? evp_keymgmt_importdomparams(keymgmt, params)
+        : evp_keymgmt_importkey(keymgmt, params);
```
*Side note*: Having enforced a separation between provider side domain parameter objects and key objects, I can't help wondering whether there should actually be a provider side distinction between public keys and private keys....anyway  not for this PR!

levitte replied:
I did think about that distinction a while ago, but considering the private key *always* contains the public components as well, it seems futile and counter-productive.

levitte replied:
Basically, one can say that we always deal with key *pairs* but allow for the private parts to be absent.

mattcaswell replied:
> the private key always contains the public components 

One could say the same thing about the public components and the domain parameters, i.e. "the public components always contain the domain parameters"

mattcaswell commented on crypto/evp/keymgmt_lib.c:
```diff
@@ -101,11 +106,11 @@ void *evp_keymgmt_export_to_provider(EVP_PKEY *pk, EVP_KEYMGMT *keymgmt)
 
-        /* Otherwise, simply use it */
-        provkey = pk->ameth->export_to(pk, keymgmt);
+        /* Otherwise, simply use it.  The export_to will tell us what it is */
```
This comment seems out of date

levitte replied:
The first sentence is still valid, but I've removed the second sentence.

mattcaswell reviewed (approved):
Approved assuming the comment is fixed

levitte commented:
Thanks for the review!

Merged.

02f060d17e667a2805eb0c71266c35de9e7e3864 PKEY: adapt the export_to_provider funtions to handle domain params too
14e3e00fe2c20a8594e3e20545d9f001fd7fa850 DH: Add export of domain parameters to provider
073f59c407b06c1b64d84808f1bee9f9457222f9 DSA: Add export of domain parameters to provider
c8f2301629b06ef43767d7d2750afaadc3d55deb DH: Add export of keys and domain parameters from provider
13aa5d29601683e0971763836ec37302fc7cece9 DSA: Add export of keys and domain parameters from provider
cd32a0f5894344b6c8739a3586a20683a6bf2d5a Don't abuse the API when that's not what is tested

romen commented on include/crypto/asn1.h:
```diff
@@ -70,7 +70,8 @@ struct evp_pkey_asn1_method_st {
     /* Exports to providers */
     size_t (*dirty_cnt) (const EVP_PKEY *pk);
-    void *(*export_to) (const EVP_PKEY *pk, EVP_KEYMGMT *keymgmt);
+    void *(*export_to) (const EVP_PKEY *pk, EVP_KEYMGMT *keymgmt,
```
@levitte sorry to come late with this comment, but would it be possible to document the expected behaviour of an `export_to` function?

I know it's an internal function, but being an internal "interface" it warrants a bit of care even with things still changing (or especially because of this).

I encountered this while working on #10631 and noticed only now that both `dirty_cnt` and `export_to` are underdocumented and the only option right now is to `git blame` to see in which PR they were added, read through the discussions, and reverse engineer what the `dh` and `dsa` implementations currently do.

levitte replied:
It's there.  `perldoc -o man doc/man3/EVP_PKEY_ASN1_METHOD.pod`, then scroll down a bit.
